google/cloud/pubsub/internal/subscription_session_test.cc

```diff
@@ -101,6 +102,74 @@ TEST(SubscriptionSessionTest, ScheduleCallbacks) {
   for (auto& t : tasks) t.join();
 }
 
+/// @test Verify callbacks are scheduled in sequence.
+TEST(SubscriptionSessionTest, SequencedCallbacks) {
+  auto mock = std::make_shared<pubsub_testing::MockSubscriberStub>();
+  pubsub::Subscription const subscription("test-project", "test-subscription");
+
+  std::mutex mu;
+  int count = 0;
+  auto generate_3 = [&](google::cloud::CompletionQueue&,
+                        std::unique_ptr<grpc::ClientContext>,
+                        google::pubsub::v1::PullRequest const& request) {
+    EXPECT_EQ(subscription.FullName(), request.subscription());
+    google::pubsub::v1::PullResponse response;
+    for (int i = 0; i != 3; ++i) {
+      auto& m = *response.add_received_messages();
+      std::lock_guard<std::mutex> lk(mu);
+      m.set_ack_id("test-ack-id-" + std::to_string(count));
+      m.mutable_message()->set_message_id("test-message-id-" +
+                                          std::to_string(count));
+      ++count;
+    }
+    return make_ready_future(make_status_or(response));
+  };
+  EXPECT_CALL(*mock, AsyncPull(_, _, _)).WillOnce(generate_3);
+
+  // This receives a message, verify that it is the next message in the sequence
+  // and then sets up the expectations for the next set of calls (mostly
+  // AsyncPull).
+  std::atomic<int> expected_message_id{0};
+  auto handler = [&](pubsub::Message const& m, pubsub::AckHandler h) {
```

Comment:
This seems to imply that `handler` will be called after the first mock call, which would break the rule that expectations must be set before the first mock call.

Reply:
> This seems to imply that `handler` will be called after the first mock call,

Necessarily so: the values that the handler is called with are generated by the `AsyncPull()` mock.

> which would break the rule that expectations must be set before the first mock call.

What I am trying to test is that once values are consumed (and only then) we expect another mock call to be made, and only once those are consumed we expect the next mock call, and so on and so forth.

Let me take another stab at expressing this with `::testing::InSequence` or something...

Reply:
PTAL.